cpp/include/cudf/detail/iterator.cuh

```diff
@@ -0,0 +1,177 @@
+ * @brief value accessor of column without null bitmask
+ * A unary functor returns scalar value at `id`.
+ * `operator() (cudf::size_type id)` computes `element`
+ * This functor is only allowed for non-nullable columns.
+ *
+ * the return value for element `i` will return `column[i]`
+ *
+ * @throws `cudf::logic_error` if the column is nullable.
+ *
+ * @tparam Element The type of elements in the column
+ * -------------------------------------------------------------------------**/
+
+template <typename Element>
+struct no_null_value_accessor
+{
+  column_device_view const col;         ///< column view of column in device
+
+/** -------------------------------------------------------------------------*
+ * @brief constructor
+ * @param[in] _col column device view of cudf column
+ * -------------------------------------------------------------------------**/
+  no_null_value_accessor(column_device_view const& _col)
+    : col{_col}
+  {
+    // verify valid is null
+    CUDF_EXPECTS(!_col.nullable(), "Unexpected nullable column.");
+  }
+
+  CUDA_DEVICE_CALLABLE
+  Element operator()(cudf::size_type i) const {
+    return col.element<Element>(i);
+  }
+};
+
+
+/**
+ * @brief Constructs an iterator over a column's values that replaces null
+ * elements with a specified value.
+ *
+ * Dereferencing the returned iterator for element `i` will return `column[i]`
+ * if it is valid, or `null_replacement` if it is null.
+ * This iterator is only allowed for nullable columns.
+ *
+ * @throws `cudf::logic_error` if the column is not nullable.
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @param column The column to iterate
+ * @param null_replacement The value to return for null elements
+ * @return auto Iterator that returns valid column elements, or a null
+ * replacement value for null elements.
+ */
+template <typename Element>
+auto make_null_replacement_iterator(column_device_view const& column,
+                                    Element const null_replacement = Element{0})
+{
+  CUDF_EXPECTS(data_type(experimental::type_to_id<Element>()) == column.type(), "the data type mismatch");
+
+  return thrust::make_transform_iterator(
+      thrust::counting_iterator<cudf::size_type>{0},
+      value_accessor<Element>{column, null_replacement});
+}
+
+
+/**
+ * @brief Constructs an iterator over a column's values
+ *
+ * Dereferencing the returned iterator for element `i` will return `column[i]`
+ * This iterator is only allowed for non-nullable columns.
+ *
+ * @throws `cudf::logic_error` if the column is nullable.
+ * @throws `cudf::logic_error` if column datatype and Element type mismatch.
+ *
+ * @tparam Element The type of elements in the column
+ * @param column The column to iterate
+ * @return auto Iterator that returns valid column elements
+ */
+template <typename Element>
+auto make_no_null_iterator(column_device_view const& column)
```

Comment:
This looks to be unnecessary. If the column doesn't contain null values, then you can just call `col.begin<Element>()`. No need for a special `make_*_iterator`.

Reply:
`col.begin<Element>()` is not present for column_device_view, only present for column_view. Also, `col.begin<string_view>()` won't work, So, it has to access through `col.element<string_view>(index)`. That's the reason for creating this functor.

Reply:
Should I keep or delete `make_no_null_iterator`?

Reply:
Delete.

Reply:
deleted.